eth/backend.go

```diff
@@ -330,6 +330,38 @@ func (s *Ethereum) Etherbase() (eb common.Address, err error) {
 	return common.Address{}, fmt.Errorf("etherbase must be explicitly specified")
 }
 
+// isLocalAccount checks whether the specified address is a local account.
+//
+// This function is used during block chain reorg checking to determine
+// whether a block is mined by local accounts. We regard three types of
+// accounts as local account: etherbase, accounts in keystore or accounts
+// specified via `txpool.locals` flag.
+func (s *Ethereum) isLocalAccount(addr common.Address) bool {
+	s.lock.RLock()
+	etherbase := s.etherbase
+	s.lock.RUnlock()
+	// Check whether the given address is etherbase.
+	if addr == etherbase {
+		return true
+	}
+	// Check whether the given address in the keystore.
+	if wallets := s.AccountManager().Wallets(); len(wallets) > 0 {
+		for _, account := range wallets[0].Accounts() {
+			if account.Address == addr {
+				return true
+			}
+		}
+	}
```

Comment:
I'm a bit reluctant about iterating over the account lists. This is generally an expensive operation. The transaction pool also did something similar and we had to remove it due to exchanges and other power users having hundreds of thousands of accounts, which we can't afford to iterate.

For the purpose of this PR (i.e. miners) I think it's fine to only check the etherbase and the txpool.locals. Those are the ones that are important anyway. For average nodes, this optimization is moot anyway since they won't mine nodes themselves.